Add Crumbling Attuned Shard

edralzar opened the pull request:
This variant of an Attuned Shard is less costly to create in
a warp-plate, but can only be used once. For the same target
warp-plate it stacks up to 16 (and one can create the full
stack at once when attuning).

Each separate teleport event will consume 1 Crumbling Attuned
Shard when teleporting to the associated destination.

Crumbling Attuned Shards are currently prioritized when
selecting the destination (as opposed to default random picking
of a shard).

edralzar committed:
Add Crumbling Attuned Shard

This variant of an Attuned Shard is less costly to create in
a warp-plate, but can only be used once. For the same target
warp-plate it stacks up to 16 (and one can create the full
stack at once when attuning).

Each separate teleport event will consume 1 Crumbling Attuned
Shard when teleporting to the associated destination.

Crumbling Attuned Shards are currently prioritized when
selecting the destination (as opposed to default random picking
of a shard).

edralzar committed:
simplify the code/recipe, keep old selection method

edralzar committed:
stack size 4, create 4 shards out of 1 pearl

simonbasle commented:
simplified the code quite a bit, rebalanced the "recipe" for `Crumbling Attuned Shard` and removed the part where it would take priority when selecting a target (now random, unless there is quartz => round robin).

I'm planning on adding a scroll variant to address the case where one wants to bring multiple people to a secret place without altering the original shard selection in a warp plate.

edralzar committed:
Crumbling shard useable, fix roundrobin, centralize destroy

The crumbling shard can be used out of a warp plate
entirely, like a scroll, except it incurs the same cost
as teleporting from a plate.

The destroying of used shards is centralized in
PlayerWaystoneManager, as the plate now explicitly
sets the context.warpItem when triggering the tp.

Also fix the warp plate's round robin / tracking of
last used slot to accommodate crumbling shards being
destroyed on teleport (which could skew the index).

edralzar committed:
Extend ScrollBaseItem to benefit from scroll use fx

BlayTheNinth commented on shared/src/main/java/net/blay09/mods/waystones/block/entity/WarpPlateBlockEntity.java:
```diff
@@ -350,31 +368,70 @@ && getItem(3).getItem() == ModItems.warpDust
+
     @Nullable
-    public IWaystone getTargetWaystone() {
+    public Pair<IAttunementItem, ItemStack> getTargetAttunement() {
```
Since `IAttunementItem` is always `ItemStack.getItem()`, it would be better to just return the ItemStack only to avoid the potential for inconsistencies.

BlayTheNinth commented on shared/src/main/java/net/blay09/mods/waystones/block/entity/WarpPlateBlockEntity.java:
```diff
@@ -350,31 +368,70 @@ && getItem(3).getItem() == ModItems.warpDust
-    public IWaystone getTargetWaystone() {
+    public Pair<IAttunementItem, ItemStack> getTargetAttunement() {
         boolean useRoundRobin = false;
-        List<ItemStack> attunedShards = new ArrayList<>();
```
These changes don't seem to be necessary

edralzar replied:
maybe not, I need to think a bit more about that. IIRC I initiated the change because the current tracking of `lastAttunementSlot` for round-robin purposes falls apart when a stack of crumbling attuned shards is emptied.

BlayTheNinth commented on shared/src/main/java/net/blay09/mods/waystones/block/entity/WarpPlateBlockEntity.java:
```diff
@@ -350,31 +368,70 @@ && getItem(3).getItem() == ModItems.warpDust
     }
 
+    @Nullable
+    public IWaystone getTargetWaystone(@Nullable Pair<IAttunementItem, ItemStack> targetAttunement) {
```
This is the same as `WaystonesAPI.getBoundWaystone()`

edralzar replied:
I'm about to fix my PR to ensure I don't call `WaystonesAPI.getBoundWaystone(ItemStack is)`  with a null parameter, but wdyt about making that API method accept `@Nullable ItemStack` anyway?

BlayTheNinth replied:
ItemStack should never be nullable since there is ItemStack.EMPTY

edralzar replied:
okay, fixed

BlayTheNinth commented on shared/src/main/java/net/blay09/mods/waystones/item/CrumblingAttunedShardItem.java:
```diff
@@ -0,0 +1,151 @@
+import java.util.List;
+import java.util.Objects;
+
+public class CrumblingAttunedShardItem extends ScrollItemBase implements IAttunementItem, IFOVOnUse, IResetUseOnDamage {
```
These should not be usable like scrolls imo - we already have Bound Scrolls and having those be bindable to a warp plate (right click or by placing inside) would be much more consistent design-wise

edralzar replied:
I removed that aspect, I agree it's causing too much confusion with scrolls.

but that leaves a gap for my desired use case: I'd like to use crumbling attuned shards as a quick and dirty way to temporarily re-route _any_ warp plate towards a secret destination, without having to come back to the plate at all (be it to re-insert the original shard or remove the "secret destination" shard).

you were concerned that the randomness aspect would be missed in adventure scenarios, and maybe a good compromise would be to simply ignore full shards when crumbling shards are present? It would basically default to a random pick among the crumbling attuned shards only in that case. wdyt?

BlayTheNinth replied:
Maybe right-clicking a warp stone with a crumbling shard could consume it to "charge" the warp plate for one teleport to that destination? It could go along with some special particles forming at its base or something. Though I'd recommend splitting that out into a separate PR to keep things focused to one feature at a time

edralzar replied:
I've started working on that in a separate branch per your suggestion, but I feel there's significantly more work there. plus it adds a secondary concept / interaction which also makes it kinda confusing 🤔 

in the meantime, I've pushed a commit that reverts to the old picking algorithm but remove non-crumbling shards from the list if at least one crumbling shard is present.

as a result, one can put several stacks from several destinations and they'll get picked up randomly (or round-robin) until no more crumbling attuned shards are present, then it's back to the normal shards.

edralzar replied:
![image](https://github.com/TwelveIterationMods/Waystones/assets/2642110/bf4028fa-c776-469e-ac60-0dafc17f4661)
in the above case, supposing each shard stack is a different destination, then the first 5 teleports will randomly pick between 1c, 2c and 3c, then further teleports will randomly pick between 0 and 4.

BlayTheNinth commented on shared/src/main/java/net/blay09/mods/waystones/core/WarpMode.java:
```diff
@@ -17,6 +17,7 @@ public enum WarpMode {
     WAYSTONE_TO_WAYSTONE(() -> WaystonesConfig.getActive().xpCost.waystoneXpCostMultiplier, WarpMode::waystoneIsActivated, false),
     SHARESTONE_TO_SHARESTONE(() -> WaystonesConfig.getActive().xpCost.sharestoneXpCostMultiplier, WarpMode::sharestonesOnly, false),
     WARP_PLATE(() -> WaystonesConfig.getActive().xpCost.warpPlateXpCostMultiplier, WarpMode::warpPlatesOnly, false),
+    WARP_PLATE_CONSUMES(() -> WaystonesConfig.getActive().xpCost.warpPlateXpCostMultiplier, WarpMode::warpPlatesOnly, true),
```
I would prefer not adding another warp mode as I intend to move away from using this class at all - since the warp plate is in charge of the teleport, it can take care of removing the item on its own

edralzar replied:
done. may I ask what do you have in mind in replacement for that enum? I was actually wondering about the consume an item aspect (which could have been driven by an interface instead for example)

BlayTheNinth replied:
The `WaystoneTeleportContext` was added fairly recently and eventually I would see the different warp mode specialties within fields of that class instead.

BlayTheNinth commented on shared/src/main/java/net/blay09/mods/waystones/block/entity/WarpPlateBlockEntity.java:
```diff
@@ -281,9 +297,11 @@ private int getWarpPlateUseTime() {
-        PlayerWaystoneManager.tryTeleportToWaystone(entity, targetWaystone, WarpMode.WARP_PLATE, getWaystone())
-                .ifLeft(entities -> entities.forEach(this::applyWarpPlateEffects));
+    private boolean teleportToWarpPlate(Entity entity, IWaystone targetWaystone, Pair<IAttunementItem, ItemStack> targetAttunement) {
+        WarpMode mode = (targetAttunement.getFirst() == ModItems.crumblingAttunedShard) ? WarpMode.WARP_PLATE_CONSUMES : WarpMode.WARP_PLATE;
```
This would benefit from a tag instead of direct item comparison

edralzar committed:
review: deal with ItemStack, add tag for consumable shard(s)

edralzar committed:
review: remove the warpMode and useable aspect

edralzar committed:
fix: don't pass null to WaystonesAPI::getBoundWaystone

edralzar committed:
getTargetAttunementStack returns ItemStack.EMPTY, not null

edralzar committed:
remove the changes to round-robin algo, but prioritize crumbling shards

edralzar committed:
Merge branch 'TwelveIterationMods:1.20.x' into crumbling-attuned-shard

edralzar committed:
Fix tooltip texts, simplify CrumblingAttunedShardItem with extends

edralzar committed:
Opt-in to crumbling shards prioritization by putting a spider eye

BlayTheNinth commented:
> if https://github.com/TwelveIterationMods/Waystones/pull/746 is merged before this one, this should include a modification of the logic that limits central stack size to 1 (as crumbling attuned shards stack up to 4).

I believe that should make no difference since the limitation only prevents putting more than one item, not taking more than one
